go/vt/vtgate/executor_dml_test.go

```diff
@@ -388,6 +388,202 @@ func TestInsertSharded(t *testing.T) {
+		BindVariables: map[string]*querypb.BindVariable{
+			"music_id": sqltypes.Int64BindVariable(2),
+		},
+	}, {
+		Sql: "select user_id from music_user_map where music_id = :music_id",
+		BindVariables: map[string]*querypb.BindVariable{
+			"music_id": sqltypes.Int64BindVariable(3),
+		},
+	}, {
+		Sql: "select user_id from music_user_map where music_id = :music_id",
+		BindVariables: map[string]*querypb.BindVariable{
+			"music_id": sqltypes.Int64BindVariable(4),
+		},
+	}, {
+		Sql: "select user_id from music_user_map where music_id = :music_id",
+		BindVariables: map[string]*querypb.BindVariable{
+			"music_id": sqltypes.Int64BindVariable(5),
+		},
+	}, {
+		Sql: "select user_id from music_user_map where music_id = :music_id",
+		BindVariables: map[string]*querypb.BindVariable{
+			"music_id": sqltypes.Int64BindVariable(6),
+		},
+	}, {
+		Sql: "insert ignore into ins_lookup(fromcol, tocol) values (:fromcol0, :tocol0), (:fromcol1, :tocol1), (:fromcol2, :tocol2), (:fromcol3, :tocol3), (:fromcol4, :tocol4)",
+		BindVariables: map[string]*querypb.BindVariable{
+			"fromcol0": sqltypes.Int64BindVariable(1),
+			"tocol0":   sqltypes.Uint64BindVariable(1),
+			"fromcol1": sqltypes.Int64BindVariable(3),
+			"tocol1":   sqltypes.Uint64BindVariable(1),
+			"fromcol2": sqltypes.Int64BindVariable(4),
+			"tocol2":   sqltypes.Uint64BindVariable(1),
+			"fromcol3": sqltypes.Int64BindVariable(5),
+			"tocol3":   sqltypes.Uint64BindVariable(1),
+			"fromcol4": sqltypes.Int64BindVariable(6),
+			"tocol4":   sqltypes.Uint64BindVariable(3),
+		},
+	}, {
+		Sql: "select fromcol from ins_lookup where fromcol = :fromcol and tocol = :tocol",
+		BindVariables: map[string]*querypb.BindVariable{
+			"fromcol": sqltypes.Int64BindVariable(1),
+			"tocol":   sqltypes.Uint64BindVariable(1),
+		},
+	}, {
+		Sql: "select fromcol from ins_lookup where fromcol = :fromcol and tocol = :tocol",
+		BindVariables: map[string]*querypb.BindVariable{
+			"fromcol": sqltypes.Int64BindVariable(3),
+			"tocol":   sqltypes.Uint64BindVariable(1),
+		},
+	}, {
+		Sql: "select fromcol from ins_lookup where fromcol = :fromcol and tocol = :tocol",
+		BindVariables: map[string]*querypb.BindVariable{
+			"fromcol": sqltypes.Int64BindVariable(4),
+			"tocol":   sqltypes.Uint64BindVariable(1),
+		},
+	}, {
+		Sql: "select fromcol from ins_lookup where fromcol = :fromcol and tocol = :tocol",
+		BindVariables: map[string]*querypb.BindVariable{
+			"fromcol": sqltypes.Int64BindVariable(5),
+			"tocol":   sqltypes.Uint64BindVariable(1),
+		},
+	}, {
+		Sql: "select fromcol from ins_lookup where fromcol = :fromcol and tocol = :tocol",
+		BindVariables: map[string]*querypb.BindVariable{
+			"fromcol": sqltypes.Int64BindVariable(6),
+			"tocol":   sqltypes.Uint64BindVariable(3),
+		},
+	}}
+	if !reflect.DeepEqual(sbclookup.Queries, wantQueries) {
+		t.Errorf("sbclookup.Queries: \n%+v, want \n%+v", sbclookup.Queries, wantQueries)
+	}
+
+	// Test the 0 rows case,
```

Comment:
@sougou -- all your points make sense and I definitely appreciate the complexity here. However IMO the proposed behavior, although improved, still doesn't provide the proer semantics in various cases that I expect we'll run into given our current usage, both for ignore and upsert.

First for the ignore case, we currently use that query pattern when bulk migrating data from our existing mysql dbs, and expect that multiple insert ignore statements are idempotent and can be used to repair potentially failed operations. For example, given a table with a secondary vindex, suppose that when running an insert ignore, the entry gets added to the secondary vindex but fails when inserting into the main table, either due to a temporary network blip, a bug in the code, or any other reason. In vanilla mysql, I could detect the discrepancy and safely repair it by just re-running the insert ignore over and over again, but in the currently proposed behavior it would never insert the row into the main table due to the entry in the vindex table.

Furthermore, in your self-described (bad) example, if the application really wanted to ensure uniqueness for both columns, then the right way to do that is to have a mysql unique index on each column of the table, in addition to the two unique vindexes. Then even in the example you proposed, the insert of (1,2) would fail to add a row when applied to the actual table just like it did when attempted on the vindex table. It doesn't seem to me like a big limitation if vindexes aren't reliable as an enforcement mechanism for uniqueness on columns in the actual table.

In the upsert case, as you point out it is even more intricate, but I again respectfully disagree with your conclusion :sunglasses: We make heavy use of upserts within the Slack app, and depend on being able to update non-PK fields if rows already exist. With the currently proposed semantics, the presence of the vindex row breaks this guarantee, since if the row exists in the original table, then the vindex entry will also exist, and as a result the update portion of the upsert won't actually be executed which violates the expected semantics.

Bubbling this up to a higher level point, it seems to me that vitess is doing too much in both these cases to protect the user from their own actions. Yes, there is certainly a risk that if used incorrectly in the ways that you point out, it is possible that the coupling between the vindex and the original row becomes out of sync. However, I think it's more confusing that vitess might end up not executing the query on the table that was actually given by the user in the original query _at all_, based on the presence of a conflicting row in the vindex table.